### PR TITLE
Pass compiler options to preprocessor

### PR DESCRIPTION
**Pull request by llvm-beanz**
If we don't pass the compiler options to the preprocessor the
preprocessor doesn't get initialized correctly.

Without this change the preprocessor step identifies the shader model
version as 0.0, pipeline stage as invalid, and language version as the
default.

### CODE REVIEW

File: tools/clang/tools/dxclib/dxc.cpp [hunk 1]
```diff
@@ -924,7 +924,14 @@ void DxcContext::Preprocess() {
...
   CComPtr<IDxcOperationResult> pPreprocessResult;
   CComPtr<IDxcBlobEncoding> pSource;
+  
+  std::vector<std::wstring> argStrings;
+  CopyArgsToWStrings(m_Opts.Args, CoreOption, argStrings);
+
   std::vector<LPCWSTR> args;
+  args.reserve(argStrings.size());
+  for (const std::wstring &a : argStrings)
+    args.push_back(a.data());
```
Comment (pow2clk) [inline tools/clang/tools/dxclib/dxc.cpp:934]: There are a couple places below where `m_Opts.Args` are explicitly added to the preprocessor `args`. They would now add duplicates.

**Review by pow2clk — COMMENTED** [review]
I'd like to avoid the duplicate arguments.

I can't come up with any reason why including all the arguments for the preprocessor would cause a problem, but I also don't know why they were excluded in the first place. Maybe @Tex3d does?

**Comment by llvm-beanz** [comment]
I was also curious if there was a reason for not passing all arguments through. I think if there is, we should filter out specific arguments rather than only passing along specific ones.

Running the preprocessor with different arguments than the compile call can (and does) result in unexpected behavior differences when preprocessing as a distinct step.

Having an accurate preprocess-only step is a precursor to our compiler being used by object caching and compilation distribution tools like ccache, sccache, distcc and incredibuild.

This also might be related to the issue reported in #4146, which I closed because the comment made it sound like user error that was corrected.

**Comment by pow2clk** [comment]
I was trying to find #4146, but couldn't because it got closed. I don't want to give you busy work, but I wonder if we should add a test using the API interface as well

**Comment by llvm-beanz** [comment]
> I was trying to find #4146, but couldn't because it got closed. I don't want to give you busy work, but I wonder if we should add a test using the API interface as well

If you think it is worth the extra test I can add it, but since DXC calls the same `IDxcCompiler::Preprocess` method, I wasn't sure it was worth it.

**Comment by llvm-beanz** [comment]
Discussed with Tex offline, he was unsure of any reason why we needed to filter the options for the preprocessor. If something pops up we can adjust this, but let's just go with this for now :)

**Pull request merged**